src/persist/src/operators/stream.rs

```diff
@@ -301,17 +329,183 @@ where
 
         (data_output_stream, error_output_stream)
     }
+
+    fn conditional_seal<K2, V2>(
+        &self,
+        name: &str,
+        condition_input: &Stream<G, ((K2, V2), u64, isize)>,
+        primary_write: StreamWriteHandle<K, V>,
+        condition_write: StreamWriteHandle<K2, V2>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    )
+    where
+        K2: TimelyData + Codec,
+        V2: TimelyData + Codec,
+    {
+        let operator_name = format!("conditional_seal({})", name);
+        let mut seal_op = OperatorBuilder::new(operator_name.clone(), self.scope());
```

Comment:
instead of using `OperatorBuilder` do you think it would be possible to use `binary_notify` for this operator? I'm assuming you've already thought about this and there's some reason, probably tied to `notificator`s that I'm missing.

Reply:
It would be possible to use `binary_notify`, yes. The reason I'm using `OperatorBuilder` is that I'm preparing for an optimization that we have to do very soon: right now, we block on the `seal()` by calling `recv()`. However, this blocks this one operator for quite a bit, making the system less responsive and not allowing as much throughput as we could.

We have to hand-roll our own little async runtime with an `ArcWake` that is backed by a `SyncActivator`, which we can only get when using the low-level operator API. This will then allow us to convert our `Futures` into "real" `Futures` and  wait for them to be ready asynchronously. (It's not terribly hard, and there is already similar code in timely: https://github.com/TimelyDataflow/timely-dataflow/blob/c5195eb566456a4cb79112fa39944490c9c00a4b/timely/src/dataflow/operators/to_stream.rs#L118)

cc @danhhz on the async stuff. And we would then also do a similar thing for writes to make the whole thing fully pipelined.